modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala

```diff
@@ -351,8 +358,13 @@ class DerivationMacros(val c: blackbox.Context) extends ScalaVersionCompat {
           }
         }
       case _root_.scala.Some(typeFieldName) =>
-        _root_.io.circe.Decoder[_root_.java.lang.String].prepare(_.downField(typeFieldName)).flatMap {
-          case ..$discriminatorCases
+        new _root_.io.circe.Decoder[$tpe] {
+          // NOTE: This wrapper might look pointless, but leaving it out causes compilation to fail with
+          // an assertion error from the compiler. There's probably a better way to avoid the problem.
```

Comment:
This is the problem I mentioned in the ticket. I noticed that the compiler crash only happened in the `Decoder` case, not the `Codec`. The only difference I could see was that the `Codec` explicitly created a wrapper class around the generated `case` statements, so I tried doing that and it worked. This is my first time digging into Scala macros so I have no idea why the other way wouldn't work.